src/auth/downscopedclient.ts

```diff
@@ -232,7 +246,65 @@ export class DownscopedClient extends AuthClient {
     opts: GaxiosOptions,
     callback?: BodyResponseCallback<T>
   ): GaxiosPromise<T> | void {
-    throw new Error('Not implemented.');
+    if (callback) {
+      this.requestAsync<T>(opts).then(
+        r => callback(null, r),
+        e => {
+          return callback(e, e.response);
+        }
+      );
+    } else {
+      return this.requestAsync<T>(opts);
+    }
+  }
+
+  /**
+   * Authenticates the provided HTTP request, processes it and resolves with the
+   * returned response.
+   * @param opts The HTTP request options.
+   * @param retry Whether the current attempt is a retry after a failed attempt.
+   * @return A promise that resolves with the successful response.
+   */
+  protected async requestAsync<T>(
+    opts: GaxiosOptions,
+    retry = false
+  ): Promise<GaxiosResponse<T>> {
```

Comment:
Is this logic identical to `BaseExternalClient`? If so I'm wondering if we could figure out a way to extract this into a mixin:

https://www.typescriptlang.org/docs/handbook/mixins.html

One option could be even doing this:

```js
protected async requestAsync() {
  BaseExternalClient.prototype.requestAsync.apply(opts, retry);
}
```

I haven't played with this approach in TypeScript only JavaScript, but seems like it should work.

Reply:
I totally agree with the idea of reducing common code & duplication. 
The reason I'm copying the implementation is that @bojeil-google and I discussed this way is easier solution.
I just researched on both approaches, I think Mixin approach works for Javascript(I used that before), but for Typescript it looks different that could not be easily replicated in this scenario. The second approach doesn't work because some APIs like requestAsync are not public.

From my perspective, there is a good way to resolve the code duplication by creating a new class inheriting from AuthClient serving as parent class(I mentioned in design [doc](https://docs.google.com/document/d/1MJb9SH7cgGQYTUb6cShtMoOuaU3b05XtU4BVijGVok0/edit?resourcekey=0-K6R_URgisKyatWB6cXPK5g)), for both BaseExternalAccountOptions and DownscopedClient but that requires much effort which might affect timeline for the launch.

Is it possible for us to proceed with this short term solution and optimize on the replication in the future after the launch? (Since these copied APIs are there but not real use cases for developers)

Reply:
Yes, I'm open to this in the short term, but could you please open a tracking issue to loop back and fix 👍

Reply:
http://b/198674048